cpp/util/ManagedArray.h

```diff
@@ -0,0 +1,305 @@
+ *  underlying data has been destructed by the owner. This is generally safe
+ *  within the freud C++ API because compute classes in freud are designed to
+ *  accept ManagedArrays as inputs, which therefore must be constructed by the
+ *  calling code. Within the freud Python API, this logic is handled through
+ *  proper usage of the acquisition methods of this class, which allow
+ *  ManagedArrays to transfer data ownership. The class is designed around the
+ *  expectation of a Python wrapper class that will maintain ownership of a
+ *  ManagedArray and its underlying data between compute calls.
+ *
+ *  There are two specific scenarios when ManagedArrays are invalidated.
+ *      1. ManagedArrays in state 2 will be invalidated if the data source
+ *         array is deleted.
+ *      2. ManagedArrays in state 3 will be invalidated if the data source is
+ *         deleted or changes in size.
+ *
+ *  No attempt is made to protect against these failures. The user is
+ *  responsible for avoiding these cases.
+ */
+template<typename T> class ManagedArray
+{
+public:
+    //! Default constructor.
+    /*! For safety, these arrays must be constructed with the manage argument
+     *  set to true in order to create an empty instance that manages its own
+     *  memory. By default, constructed arrays do not manage their own memory to
+     *  avoid accidental memory leaks.
+     *
+     *  \param manage Whether or not this instance should manage its own memory
+     *                (Default: false).
+     */
+    ManagedArray(bool manage=false) : m_size(nullptr), m_data(nullptr), m_managed(manage), m_external_ptr(false) {}
+
+    //! Constructor with specific size for thread local arrays
+    /*! When using this constructor, the class automatically manages its own
+     *  memory since it is allocating it.
+     *
+     *  \param size Size of the array to allocate.
+     */
+    ManagedArray(unsigned int size) : m_managed(true), m_external_ptr(false)
+    {
+        m_size = new unsigned int(size);
+        m_data = new T[size];
+        reset();
+    }
+
+    //! Construct object from existing array.
+    /*! \param T* Pointer to existing data.
+     *  \param size Size of the array to allocate.
+     */
+    ManagedArray(T* array, unsigned int size) : m_managed(false), m_external_ptr(true)
+    {
+        m_size = new unsigned int(size);
+        m_data = array;
+    }
+
+    //! Copy constructor.
+    /*! This constructor is required to ensure that the original object always
+     *  owns its own memory. However ManagedArrays pointing to external data will always store
+     *
+     *  \param other ManagedArray instance to copy.
+     */
+    ManagedArray(const ManagedArray &other) : m_data(other.m_data), m_managed(false), m_external_ptr(other.m_external_ptr)
+    {
+        if (m_external_ptr)
+        {
+            m_size = new unsigned int(*other.m_size);
+        }
+        else
+        {
+            m_size = other.m_size;
```

Comment:
This leaves the implementation open to a segfault if you make a copy of a parent object that later goes out of scope (when it deallocates its `m_size`), right?

Reply:
Correct. As noted in the C++ class documentation, there are currently two cases where a ManagedArray can be invalidated: 
```
There are two specific scenarios when ManagedArrays are invalidated.
    1. ManagedArrays in state 2 will be invalidated if the data source
       array is deleted.
    2. ManagedArrays in state 3 will be invalidated if the data source is
        deleted or changes in size.
```

That's true not just for deallocating `m_size`, but also the underlying `m_data` pointer to the array itself. If we do what I suggested in my longer comment and stop using this object as a pointer to NumPy data, then we could write this class as pure C++ and use smart pointers to alleviate this issue. Currently the problem is that trying to point to a numpy.ndarray means that we can't use smart pointers without adding other complexity because they'll delete the data even if the numpy array should be owning it. If we stop pointing to numpy arrays some of these problems go away.